keps/sig-architecture/2019-03-19-go-modules.md

```diff
@@ -0,0 +1,248 @@
+        * pulls latest commit of transitive non-module-based dependencies (likely to break)
+      * GOPATH consumers import `k8s.io/apimachinery/...` (as they do today)? (needs research to verify)
+      * module-based consumers import `k8s.io/apimachinery/v15/...`
+      * module-based consumers can have multiple copies of kubernetes libraries (though non-module transitive dependencies could still conflict)
+    * kubernetes tooling
+      * requires rewriting all `k8s.io/{component}/vX/...` imports when a major version bump occurs
+      * requires updating code generation scripts to generate versioned imports for `k8s.io/{api,apimachinery,client-go}`, etc
+      * might require changes to build scripts to include version in built package path (needs research to verify)
+      * requires tooling to detect when a breaking go change has occurred in a particular component relative to all tagged releases for the current major version
+      * requires tooling to manage versions per component (instead of homogenous versions for staging components)
+
+### Build synthetic godeps.json files
+
+Godeps.json files are consumed by some dependency management tools (like `glide` and `dep`).
+Existing consumers of published staging components that are not yet using go modules could be impacted by removal of Godeps.json files.
+To mitigate this, we could consider generating synthetic Godeps.json files based on go.mod dependency versions.
+This section is in progress, pending investigation into necessity and feasibility.
+
+## Design Details
+
+### Test Plan
+
+* CI scripts to verify vendor contents are recreatable and match referenced versions
+* CI scripts to verify vendor licenses are up to date
+* CI scripts to verify staging component dependencies are correct
+* CI scripts to verify staging component publishing is successful
+* CI scripts to verify examples using `k8s.io/client-go` can be consumed and build automatically and successfully by GOPATH and module-based consumers
+
+### Graduation Criteria
+
+* kubernetes/kubernetes vendor management uses go modules
+* published staging components can be successfully consumed by go module consumers
+* published staging components can be successfully consumed by GOPATH-based consumers
+
+### Upgrade / Downgrade Strategy
+
+Not applicable
+
+### Version Skew Strategy
+
+Not applicable
+
+## Implementation History
+
+- 2019-03-19: Created
+
+## Alternatives
+
+* Continue using `godep` for vendor management. This is not viable for several reasons:
+  * The tool is unmaintained (the project readme states "Please use dep or another tool instead.")
+  * There are significant performance problems (the pull-kubernetes-godeps CI job takes ~30 minutes to verify vendoring)
+  * There are significant functional problems (the tool cannot be run in some environments, gets confused by diamond dependencies, etc)
+
+* Use an alternate dependency mechanism (e.g. `dep`, `glide`). This is not preferred for several reasons:
```

Comment:
As the unmaintainer of glide I'll go on the record as saying it should not be used. There's a go issue about archiving dep, as well. Last I checked this was the go teams plan.